Framework/Source/Graphics/Program/ProgramReflection.cpp

```diff
@@ -0,0 +1,1190 @@
+                case 4:
+                    return ReflectionBasicType::Type::Float3x4;
+                }
+                break;
+            case 4:
+                switch (columns)
+                {
+                case 2:
+                    return ReflectionBasicType::Type::Float4x2;
+                case 3:
+                    return ReflectionBasicType::Type::Float4x3;
+                case 4:
+                    return ReflectionBasicType::Type::Float4x4;
+                }
+                break;
+            }
+        }
+
+        should_not_get_here();
+        return ReflectionBasicType::Type(-1);
+    }
+
+    static ReflectionResourceType::StructuredType getStructuredBufferType(TypeReflection* pSlangType)
+    {
+        auto invalid = ReflectionResourceType::StructuredType::Invalid;
+
+        if (pSlangType->getKind() != TypeReflection::Kind::Resource)
+            return invalid; // not a structured buffer
+
+        if (pSlangType->getResourceShape() != SLANG_STRUCTURED_BUFFER)
+            return invalid; // not a structured buffer
+
+        switch (pSlangType->getResourceAccess())
+        {
+        default:
+            should_not_get_here();
+            return invalid;
+
+        case SLANG_RESOURCE_ACCESS_READ:
+            return ReflectionResourceType::StructuredType::Default;
+
+        case SLANG_RESOURCE_ACCESS_READ_WRITE:
+        case SLANG_RESOURCE_ACCESS_RASTER_ORDERED:
+            return ReflectionResourceType::StructuredType::Counter;
+        case SLANG_RESOURCE_ACCESS_APPEND:
+            return ReflectionResourceType::StructuredType::Append;
+        case SLANG_RESOURCE_ACCESS_CONSUME:
+            return ReflectionResourceType::StructuredType::Consume;
+        }
+    };
+
+    ReflectionVar::SharedPtr reflectVariable(VariableLayoutReflection* pSlangLayout, const ReflectionPath* pPath);
+    ReflectionType::SharedPtr reflectType(TypeLayoutReflection* pSlangType, const ReflectionPath* pPath);
+
+    // Once we've found the path from the root down to a particular leaf
+    // variable, `getDescOffset` can be used to find the final summed-up descriptor offset of the element
+    static uint32_t getDescOffsetFromPath(const ReflectionPath* pPath, uint32_t arraySize, SlangParameterCategory category)
+    {
+#ifndef FALCOR_VK
+        return 0;
+#else
+        uint32_t offset = 0;
+        bool first = true;
+        for (auto pp = pPath; pp; pp = pp->pParent)
+        {
+            if ((pp->pTypeLayout) && (pp->pTypeLayout->getKind() == TypeReflection::Kind::Array))
+            {
+                offset += (uint32_t)pp->childIndex * arraySize;
+                arraySize *= (uint32_t)pp->pTypeLayout->getElementCount();
+            }
+        }
+        return offset;
+#endif
+    }
+
+    static size_t getRegisterIndexFromPath(const ReflectionPath* pPath, SlangParameterCategory category)
+    {
+        uint32_t offset = 0;
+        for (auto pp = pPath; pp; pp = pp->pParent)
+        {
```

Comment:
The more I look at this code, the more convinced I get that Slang needs an API that does this for you, because the next developer who uses Slang is just going to have to write the exact same thing in their own codebase...